src/utils/futures.rs

```diff
@@ -119,21 +145,62 @@ impl RemoteThread {
-    /// `Result`, normally the output is returned in an `Ok` but when the remote
-    /// future is dropped because the thread is shut down or restarted for any
-    /// reason it returns `Err(oneshot::Canceled)`.
-    pub fn spawn<F, R, T>(&self, factory: F) -> impl Future<Output = Result<T, RemoteCanceled>>
+    /// The returned future resolves when the spawned future has completed execution.  The
+    /// output type of the spawned future is wrapped in a `Result`, normally the output is
+    /// returned in an `Ok`.
+    ///
+    /// When the future takes too long `SpawnError::Timeout` is returned, if the
+    /// `RemoteThread` it is running on is dropped `SpawnError::Canceled` is returned.
+    pub fn spawn<F, R, T>(
+        &self,
+        task_name: &'static str,
+        timeout: Duration,
+        factory: F,
+    ) -> impl Future<Output = Result<T, SpawnError>>
     where
         F: FnOnce() -> R + Send + 'static,
         R: Future<Output = T> + 'static,
         T: Send + 'static,
     {
+        enum ChannelMsg<T> {
+            Output(T),
+            Timeout,
+        };
         let (tx, rx) = oneshot::channel();
-        let msg = actix::msgs::Execute::new(|| -> Result<(), ()> {
-            let fut = factory().map(|output| tx.send(output).or(Err(())));
-            let fut01 = fut.boxed_local().compat();
+        let creation_time = Instant::now();
+        let msg = actix::msgs::Execute::new(move || -> Result<(), ()> {
+            let fut01 = futures01::future::lazy(move || {
+                metric!(
+                    timer("futures.wait_time") = creation_time.elapsed(),
+                    "task_name" => task_name,
+                );
+                let start_time = Instant::now();
+                factory()
+                    .then(future::ok)
+                    .boxed_local()
+                    .compat()
+                    .timeout(timeout)
+                    .then(move |r: Result<T, tokio::timer::timeout::Error<()>>| {
+                        metric!(
+                            timer("futures.done") = start_time.elapsed(),
+                            "task_name" => task_name,
+                            "status" => if r.is_ok() {"done"} else {"timeout"},
+                        );
+                        let msg = match r {
+                            Ok(o) => ChannelMsg::Output(o),
+                            Err(_) => {
+                                // Because we wrapped our T into future::ok() above to make a
+                                // TryFuture, we know the <Timeout as Future>::Error will only
+                                // occur if the error is actually because of a timeout, so we do
+                                // not need to check with .is_timer() or .is_inner().
+                                ChannelMsg::Timeout
+                            }
+                        };
+                        // TODO: Should send failures be logged?  Panic?
```

Comment:
So this is still an interesting question.  This would happen if the future which called `.spawn()` here was dropped itself and thus the receiver is no longer there.  This could realistically happen when e.g. that future itself times out and is cancelled.  So I think panic is likely the wrong thing to do.

I guess that leaves logging this?  Probably just as info, I doubt this warrants and error since it is not the cause but rather an effect.

Reply:
I went with info log